src/lib/operators/sort.cpp

```diff
@@ -33,19 +20,76 @@ const std::string& Sort::name() const {
 std::shared_ptr<AbstractOperator> Sort::_on_deep_copy(
     const std::shared_ptr<AbstractOperator>& copied_input_left,
     const std::shared_ptr<AbstractOperator>& copied_input_right) const {
-  return std::make_shared<Sort>(copied_input_left, _column_id, _order_by_mode, _output_chunk_size);
+  return std::make_shared<Sort>(copied_input_left, _sort_definitions, _output_chunk_size);
 }
 
 void Sort::_on_set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters) {}
 
 std::shared_ptr<const Table> Sort::_on_execute() {
-  _impl = make_unique_by_data_type<AbstractReadOnlyOperatorImpl, SortImpl>(
-      input_table_left()->column_data_type(_column_id), input_table_left(), _column_id, _order_by_mode,
-      _output_chunk_size);
-  return _impl->_on_execute();
+  _sort_definition_data_types.resize(_sort_definitions.size());
+  std::transform(
+      _sort_definitions.begin(), _sort_definitions.end(), _sort_definition_data_types.begin(),
+      [this](const auto sort_definition) { return input_table_left()->column_data_type(sort_definition.column); });
+
+  _validate_sort_definitions();
+
+  std::shared_ptr<const Table> output = input_table_left();
+  std::shared_ptr<Table> output_mutable;
+  std::shared_ptr<PosList> sorted_pos_list = std::shared_ptr<PosList>(nullptr);
+
+  for (auto column_id = _sort_definitions.size(); column_id-- != 0;) {
+    const bool is_last_sorting_run = (column_id == 0);
+
+    auto sort_definition = _sort_definitions[column_id];
+    const auto data_type = _sort_definition_data_types[column_id];
+
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+
+      auto sort_single_col_impl = SortImpl<ColumnDataType>(input_table_left(), sort_definition.column,
+                                                           sort_definition.order_by_mode, _output_chunk_size);
+      const auto row_id_value_vector = sort_single_col_impl.sort_one_column(sorted_pos_list);
+
+      if (is_last_sorting_run) {
+        auto materialization = std::make_shared<SortImplMaterializeOutput<ColumnDataType>>(
+            input_table_left(), row_id_value_vector, _output_chunk_size);
+
+        output_mutable = materialization->execute();
+      } else {
+        sorted_pos_list = std::make_shared<PosList>();
+
+        for (auto& row_id_value : *row_id_value_vector) {
+          sorted_pos_list->emplace_back(row_id_value.first);
+        }
+      }
+    });
+  }
+
+  auto final_sort_definition = _sort_definitions[0];
+  // Set the ordered_by attribute of the output's chunks to the column the chunk was sorted by last (because this
+  // is the column that the chunk is ordered by in any case; whereas it is not necessarily sorted by the other
+  // columns -- on their own -- as well).
+  const auto chunk_count = output_mutable->chunk_count();
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    output_mutable->get_chunk(chunk_id)->set_ordered_by(
```

Comment:
```suggestion
    const auto chunk = output->get_chunk(chunk_id);
    if (chunk->is_mutable()) {
        chunk->finalize();
    }
    chunk->set_ordered_by(
```

Hint: We used `chunk->finalize` whenever the `ordered_by` property is set to ensure that the chunk is always ordered. The performance impacts on this are still measured, though.